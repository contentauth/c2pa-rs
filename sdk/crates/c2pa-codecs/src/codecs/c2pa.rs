// Copyright 2022 Adobe. All rights reserved.
// This file is licensed to you under the Apache License,
// Version 2.0 (http://www.apache.org/licenses/LICENSE-2.0)
// or the MIT license (http://opensource.org/licenses/MIT),
// at your option.

// Unless required by applicable law or agreed to in writing,
// this software is distributed on an "AS IS" BASIS, WITHOUT
// WARRANTIES OR REPRESENTATIONS OF ANY KIND, either express or
// implied. See the LICENSE-MIT and LICENSE-APACHE files for the
// specific language governing permissions and limitations under
// each license.

use std::io::{Read, Seek, Write};

use crate::{
    ByteSpan, C2paSpan, CodecError, Decode, DefaultSpan, Embed, Embeddable, Encode, EncodeInPlace,
    Span, Support,
};

/// Supports working with ".c2pa" files containing only manifest store data
#[derive(Debug)]
pub struct C2paCodec<R> {
    src: R,
}

impl<R> C2paCodec<R> {
    pub fn new(src: R) -> Self {
        Self { src }
    }
}

impl Support for C2paCodec<()> {
    const MAX_SIGNATURE_LEN: usize = 0;

    fn supports_extension(extension: &str) -> bool {
        extension == "c2pa"
    }

    fn supports_mime(mime: &str) -> bool {
        mime == "application/c2pa" || mime == "application/x-c2pa-manifest-store"
    }
}

impl<R: Read + Seek> Decode for C2paCodec<R> {
    fn read_c2pa(&mut self) -> Result<Option<Vec<u8>>, CodecError> {
        let mut cai_data = Vec::new();
        // read the whole file
        self.src.read_to_end(&mut cai_data)?;
        Ok(Some(cai_data))
    }
}

impl<R: Read + Seek> Encode for C2paCodec<R> {
    fn write_c2pa(&mut self, mut dst: impl Write, c2pa: &[u8]) -> Result<(), CodecError> {
        // just write the store bytes and ingore the input stream
        dst.write_all(c2pa)?;
        Ok(())
    }

    fn remove_c2pa(&mut self, _dst: impl Write) -> Result<bool, CodecError> {
        // TODO: true or false?
        Ok(false)
    }
}

impl<R: Read + Write + Seek> EncodeInPlace for C2paCodec<R> {
    fn patch_c2pa(&mut self, c2pa: &[u8]) -> Result<(), CodecError> {
        self.src.write_all(c2pa)?;
        Ok(())
    }
}

impl<R: Read + Seek> Embed for C2paCodec<R> {
    fn embeddable(bytes: &[u8]) -> Result<Embeddable, CodecError> {
        Ok(Embeddable {
            bytes: bytes.to_vec(),
        })
    }

    fn embed(&mut self, embeddable: Embeddable, mut dst: impl Write) -> Result<(), CodecError> {
        dst.write_all(&embeddable.bytes)?;
        Ok(())
    }
}

impl<R: Read + Seek> Span for C2paCodec<R> {
    fn span(&mut self) -> Result<DefaultSpan, CodecError> {
        todo!()
    }

    fn c2pa_span(&mut self) -> Result<C2paSpan, CodecError> {
        Ok(C2paSpan {
            spans: vec![ByteSpan { start: 0, len: 0 }],
        })
    }
}
