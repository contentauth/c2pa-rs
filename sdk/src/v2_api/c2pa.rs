// Copyright 2024 Adobe. All rights reserved.
// This file is licensed to you under the Apache License,
// Version 2.0 (http://www.apache.org/licenses/LICENSE-2.0)
// or the MIT license (http://opensource.org/licenses/MIT),
// at your option.

// Unless required by applicable law or agreed to in writing,
// this software is distributed on an "AS IS" BASIS, WITHOUT
// WARRANTIES OR REPRESENTATIONS OF ANY KIND, either express or
// implied. See the LICENSE-MIT and LICENSE-APACHE files for the
// specific language governing permissions and limitations under
// each license.

use async_generic::async_generic;

use crate::{error::Result, Builder, CAIRead, Reader, Signer};

/// The main entry point for the v2 API
#[derive(Default)]
pub struct C2pa {
    pub verify: bool,
    pub signer: Option<Box<dyn Signer>>,
}

impl<'a> C2pa {
    /// Create a new instance of the v2 API
    pub fn new() -> Self {
        C2pa {
            verify: true,
            signer: None,
        }
    }

    /// Set the signer to use for signing
    /// # Arguments
    /// * `signer` - The signer to use
    /// # Returns
    pub fn set_signer(mut self, signer: Box<dyn Signer>) -> Self {
        self.signer = Some(signer);
        self
    }

    #[async_generic(async_signature(
        &mut self,
        format: &str,
        stream: &mut dyn CAIRead,
    ))]
    /// Create a manifest store Reader from a stream
    /// # Arguments
    /// * `format` - The format of the stream
    /// * `stream` - The stream to read from
    /// # Returns
    /// A reader for the manifest store
    /// # Errors
    /// If the stream is not a valid manifest store
    pub fn read(&self, format: &str, stream: &mut dyn CAIRead) -> Result<Reader> {
        if _sync {
            Reader::from_stream(format, stream)
        } else {
            Reader::from_stream_async(format, stream).await
        }
    }

    /// Create a new manifest builder for the v2 API
    pub fn builder(&self) -> Builder {
        Builder::new()
    }
}

#[cfg(test)]
pub(crate) mod tests {
    #![allow(clippy::expect_used)]
    #![allow(clippy::unwrap_used)]

    use std::path::{Path, PathBuf};

    use super::*;
    use crate::utils::test::{fixture_path, temp_signer};

    fn sign_and_read<P: AsRef<Path>>(c2pa: &C2pa, path: P) -> Result<()> {
        let signer = c2pa
            .signer
            .as_ref()
            .ok_or_else(|| crate::error::Error::BadParam("No signer set for C2pa".to_string()))?;
        let path = path.as_ref();
        let mut file = std::fs::File::open(path)?;
        let extension = path.extension().unwrap().to_str().unwrap();
        let target_path = PathBuf::from(env!("CARGO_MANIFEST_DIR")).join("../target");
        let dest_path = target_path.join("signed").join(path.file_name().unwrap());
        std::fs::create_dir_all(dest_path.parent().unwrap())?;
        dbg!(dest_path.to_str());
        let mut dest = std::fs::File::create(&dest_path)?;
        c2pa.builder()
            .with_json(
                "{
                \"assertions\": []
            }",
            )?
            .sign(extension, &mut file, &mut dest, signer.as_ref())
            .unwrap();
        let mut dest = std::fs::File::open(&dest_path).unwrap();
        let reader = c2pa.read(extension, &mut dest).unwrap();
        println!("{}", reader.json());
        assert!(reader.status().is_none());
        Ok(())
    }

    #[test]
    fn test_c2pa_new() {
        let signer = temp_signer();
        let c2pa = C2pa::new().set_signer(signer);
        let input_files = [
            "IMG_0003.jpg",
            "libpng-test.png",
            //"TUSCANY.tif",
            "sample1.svg",
            "sample1.webp",
            //"mars.dng",
            //"sample1.heic",
            //"sample1.heif",
            //"sample1.mp3",
            //"video1.mp4",
            //"sample1.avif",
            // "mars.mov",
            //"mars.m4a",
            //"sample1.wav",
            //"basic.pdf"
        ];
        for file in input_files.iter() {
            sign_and_read(&c2pa, fixture_path(file)).expect("Failed to sign and read");
        }
    }

    // #[test]
    // fn test_read_write_manifest_stream() {
    //     let source = crate::utils::test::fixture_path("TUSCANY.TIF");

    //     let mut source = std::fs::File::open(source).unwrap();
    //     let mut dest = std::io::Cursor::new(Vec::new()); //std::fs::File::create(&output).unwrap();
    //     let signer = temp_signer();

    //     Builder::new()
    //         .sign("application/tiff", &mut source, &mut dest, signer.as_ref())
    //         .unwrap();

    //     dest.set_position(0);
    //     // read data back
    //     let reader = Reader::from_stream("image/tiff", &mut dest).unwrap();
    //     assert!(reader.status().is_none());
    // }
}
