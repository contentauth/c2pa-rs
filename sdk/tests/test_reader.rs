// Copyright 2024 Adobe. All rights reserved.
// This file is licensed to you under the Apache License,
// Version 2.0 (http://www.apache.org/licenses/LICENSE-2.0)
// or the MIT license (http://opensource.org/licenses/MIT),
// at your option.

// Unless required by applicable law or agreed to in writing,
// this software is distributed on an "AS IS" BASIS, WITHOUT
// WARRANTIES OR REPRESENTATIONS OF ANY KIND, either express or
// implied. See the LICENSE-MIT and LICENSE-APACHE files for the
// specific language governing permissions and limitations under
// each license.

mod common;
use c2pa::{Error, Reader, Result};
use common::{assert_err, compare_to_known_good, fixture_stream};

#[test]
#[cfg(feature = "file_io")]
fn test_reader_not_found() -> Result<()> {
    let result = Reader::from_file("not_found.png");
    assert_err!(result, Err(Error::IoError(_)));
    Ok(())
}

#[test]
fn test_reader_no_jumbf() -> Result<()> {
    let (format, mut stream) = fixture_stream("sample1.png")?;
    let result = Reader::from_stream(&format, &mut stream);
    assert_err!(result, Err(Error::JumbfNotFound));
    Ok(())
}

#[test]
fn test_reader_ca_jpg() -> Result<()> {
    let (format, mut stream) = fixture_stream("CA.jpg")?;
    let reader = Reader::from_stream(&format, &mut stream)?;
    compare_to_known_good(&reader, "CA.json")
}

#[test]
fn test_reader_c_jpg() -> Result<()> {
    let (format, mut stream) = fixture_stream("C.jpg")?;
    let reader = Reader::from_stream(&format, &mut stream)?;
    compare_to_known_good(&reader, "C.json")
}

#[test]
fn test_reader_xca_jpg() -> Result<()> {
    let (format, mut stream) = fixture_stream("XCA.jpg")?;
    let reader = Reader::from_stream(&format, &mut stream)?;
    compare_to_known_good(&reader, "XCA.json")
}

#[test]
#[ignore]
/// Generates the known good for the above tests
/// This is ignored by default
/// to call use test -- --ignored
fn write_known_goods() -> Result<()> {
    let filenames = ["CA.jpg", "C.jpg", "XCA.jpg"];
    for filename in &filenames {
        common::write_known_good(filename)?;
    }
    Ok(())
}
