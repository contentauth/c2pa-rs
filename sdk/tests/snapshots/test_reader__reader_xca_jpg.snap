---
source: sdk/tests/test_reader.rs
expression: unescape_json(&reader.json())?
---
{
  "active_manifest": "[MANIFEST_URN]",
  "manifests": {
    "[MANIFEST_URN]": {
      "assertions": [
        {
          "data": {
            "@context": "http://schema.org/",
            "@type": "CreativeWork",
            "author": [
              {
                "@type": "Person",
                "name": "Adobe make_test"
              }
            ]
          },
          "kind": "Json",
          "label": "stds.schema-org.CreativeWork"
        },
        {
          "data": {
            "actions": [
              {
                "action": "c2pa.opened",
                "instanceId": "xmp.iid:813ee422-9736-4cdc-9be6-4e35ed8e41cb",
                "parameters": {
                  "ingredient": {
                    "hash": "tTBD4/E0R0AjLUdJFpsVz3lE/KJUq22Vz0UGqzhEpVs=",
                    "url": "[JUMBF_URI]"
                  }
                }
              },
              {
                "action": "c2pa.color_adjustments",
                "parameters": {
                  "name": "brightnesscontrast"
                }
              }
            ]
          },
          "label": "c2pa.actions"
        }
      ],
      "claim_generator": "make_test_images/0.24.0 c2pa-rs/0.24.0",
      "format": "image/jpeg",
      "ingredients": [
        {
          "document_id": "xmp.did:813ee422-9736-4cdc-9be6-4e35ed8e41cb",
          "format": "image/jpeg",
          "instance_id": "xmp.iid:813ee422-9736-4cdc-9be6-4e35ed8e41cb",
          "relationship": "parentOf",
          "thumbnail": {
            "format": "image/jpeg",
            "identifier": "[JUMBF_URI]"
          },
          "title": "A.jpg"
        }
      ],
      "instance_id": [XMP_ID]",
      "label": "[MANIFEST_URN]",
      "signature_info": {
        "alg": "Ps256",
        "cert_serial_number": "720724073027128164015125666832722375746636448153",
        "issuer": "C2PA Test Signing Cert",
        "time": "[TIMESTAMP1]"
      },
      "thumbnail": {
        "format": "image/jpeg",
        "identifier": "[JUMBF_URI]"
      },
      "title": "CA.jpg"
    }
  },
  "validation_status": [
    {
      "code": "assertion.dataHash.mismatch",
      "explanation": "asset hash error, name: jumbf manifest, error: hash verification( Hashes do not match )",
      "url": "[JUMBF_URI]"
    }
  ]
}
